RFC 45: Workflow (#45)
wagtail/rfcs: text/045-workflow.md

# RFC 45: Workflow

* RFC: 45
* Authors: Karl Hobley
* Created: 2019-11-12
* Last Modified: 2019-11-12

## Abstract

This RFC proposes how we can implement multi-stage configurable workflow in Wagtail core.

## Overview

### Workflows

Workflows are configured in Wagtail by an administrator. Each workflow has a name and a list of tasks arranged in a sequence.

Workflows are triggered when a user clicks the "Submit for Review" action in the page editor. This action creates a "workflow state" object and begins the first task in the workflow.

The workflow that gets used depends on the position of the page in the tree. Wagtail finds the workflow to use from the closest ancestor page that has a workflow associated with it. Out of the box, the "Root" page in Wagtail has the "Default" workflow linked with it, so pages use this workflow if none of their ancestor pages has overridden it.

The "Submit for Review" button is not shown when there isn't a workflow configured.

> ### Tasks

Author: I've been thinking of a "step" as an instance of a task in a workflow, but tasks are defined separately and I didn't think the word "step" worked very well when describing them outside of a workflow. I'll see if I can make the wording a bit clearer.

> ### Tasks

Author: Just rewrote it now. Please let me know if it's still unclear.

> ### Tasks

Reviewer: That reads so much better 👍

Tasks represent the checks that must pass for a page to be published. Each task has a name, a type, and some additional configuration dependant on the chosen type.

Tasks are defined separately to workflows, allowing workflows to share tasks.

Tasks are "triggered" when the previous task completes successfully, or when it's the first task in the workflow. Triggering a task creates a "task state" object and executes any trigger actions that the task type has defined (For example, send an email).

The "task state" object tracks the current progress for completing a task; these are associated with a workflow state and a page revision. Depending on the task type, The "task state" may be an instance of TaskState or an instance of a custom model that inherits TaskState.

> If the page is changed while a review is taking place, the existing "task state" objects will no longer be valid because they are associated with the previous revision. A new set of task state objects are created, and the review starts again at the first task.

Reviewer 2: "A new set of task state objects are created" - is a _set_ of task states really being created at this point, or would it be more accurate to say something like: a new task state corresponding to the first task in the workflow is created, and the previous active task is cancelled?

Author: Yeah, sorry the wording is a little off here. I'll fix

Author: This has been rewritten now

> If the page is changed while a review is taking place, the existing "task state" objects will no longer be valid because they are associated with the previous revision. A new set of task state objects are created, and the review starts again at the first task.

Author: Yep, this is correct. I'll mention this in the RFC

"task state" objects from previous successful reviews of the page would be available for uses such as generating diffs from the last reviewed version.

## Models

### Task

The ``Task`` model is the base that all task types inherit. It is a concrete model, so tasks use multi-table-inheritance.

This model has an ID and Name fields. It also has the following methods that are overridable:

> #### ``on_trigger(workflow_state)``

Reviewer 2: I'd suggest calling this `start`. `on_trigger` makes it sound like an event listener that accepts arbitrary callback functions, rather than a method that performs a defined bit of functionality (namely, creating / returning a TaskState).

Author: Changed

This is called when the task is triggered in a workflow. Tasks are triggered when they are the first task of the workflow, or their previous task was just completed.

This is responsible for creating an instance of ``TaskState``, which itself could be customised per task type.

#### ``get_action_menu_items(workflow_state, task_state)``

This returns a list of ``ActionMenuItem`` instances to replace the defaults. This could be used to implement "Accept" and "Reject" actions.

#### ``on_action(workflow_state, task_state, action_name)``

If the edit view receives a ``POST`` request with a field called ``wagtail-moderation-action``, then this method is called
with the action name passed through.

> #### ``user_can_access_editor(workflow_state, task_state, user)``

Reviewer: Not specifically related to this, but in a "Task 1, Task 2, Publish" workflow, will moving from one task/state to other be configurable?
i.e. editors can move to task1, task 2. but only moderators can do task 2 -> publish?

Transitions do not necessarily need page editor permissions.

Author: Yep, it's up to the developer to define who can perform these actions by overriding these methods on their task type models.

Author: Note that `user_can_access_editor` doesn't necessarily give the user permission to change content as the content can be locked by the `page_locked_for_user` method.

Returns ``True`` if the user is allowed to access the page editor. By default, this queries ``UserPagePermissionsProxy``.

> Note that this can only grant access to a user who doesn't usually have access to the editor and not take this away. To prevent users from editing pages, use ``page_locked_for_user`` instead.

Reviewer 2: Not sure I understand this comment - what's to stop people from overriding this method and defining logic that's more restrictive than the default UserPagePermissionsProxy rules?

Author: In the first iteration of this, I had it so the edit view would check the workflow then fall back to checking the `UserPagePermissionPolicy` if the workflow didn't explicitly grant edit permission to the user. I later changed it to make the workflow always check the `UserPagePermissionProxy` as this seems a little simpler implementation-wise, but it would make it possible to revoke edit privileges from someone who has it. So this comment is a little out of date but I'm still on the fence as to which of these two methods is better.

The latter method is a little simpler, but if someone uses it to revoke the editor view permission any links to the edit view will be broken because this permission policy will only be checked by the editor view and not the explorer. I'm considering changing it back, do you have any thoughts on this?

Author: It's been changed back now.

Also, this does not make the page appear in the explorer.

#### ``page_locked_for_user(workflow_state, task_state, user)``

Returns ``True`` if the page should be locked for the specified user. By default, this queries ``UserPagePermissionsProxy``.

#### ``page_lockable_by_user(workflow_state, task_state, user)``

Returns ``True`` if the page can be locked by the specified user. By default, this queries ``UserPagePermissionsProxy``.

#### ``page_unlockable_by_user(workflow_state, task_state, user)``

Returns ``True`` if the page can be unlocked by the specified user. By default, this queries ``UserPagePermissionsProxy``.

### Workflow

The ``Workflow`` model has a Name and an ordered list of tasks. It is not customisable.

> The ``Page`` model has a ``ForeignKey`` to ``Workflow``. This specifies which workflow to use on the page and its descendants.

Reviewer 2: * Do we need to consider the possibility that existing Wagtail projects might have a field named 'workflow' that would clash with this one?

* Would it be an option to use a OneToOneField here instead? (Better separation of concerns, possibly even to the point of allowing workflow to be an add-on module; less risk of unintended side effects, e.g. having to remember to ignore it when copying pages / saving revisions; more analogous to how sites / page permissions work already; possibly avoids clashing with existing fields named `workflow`, if related_name is set to `+`)

Author: Updated this to use a link table with a ``OneToOneField``

### WorkflowState and TaskState

These two models store the current status of active workflows and keep a history of all past workflows.

``WorkflowState`` has the following fields:

 - ``id``
 - ``workflow`` - A ``ForeignKey`` to ``Workflow``
 - ``page`` - A ``ForeignKey`` to ``Page``

>  - ``status`` - Either "In progress", "Approved", "Rejected" or "Cancelled"

Author: Yep, that's right. I'll mention this

 - What action buttons appear at the bottom
 - Whether the page is locked/lockable/unlockable
 - Whether the edit view is accessible to a user who hasn't got edit permission on the page (to allow non-editors to review pages)

There will also be a section in the header which shows the current workflow that is going on and what tasks have been completed.

#### Submit for moderation action button

Currently, the “Submit for moderation" button saves a revision with the “submitted for moderation” field set to True.

This is changed to use the new workflow models instead. Clicking “Submit for moderation” creates an instance of “workflow state” on the page and then triggers the first task of that workflow for the current page revision.

The workflow chosen is based on the position of the page in the tree. The “Submit for moderation” button is not displayed when there is no workflow configured for the section of the page tree or when there is a workflow currently in progress.

### The built-in workflow

> The existing “moderation” feature of Wagtail is replaced with a workflow called “Default”. The existing “submitted for moderation” field will still exist, and revisions submitted in this way will still exist in the moderation queue. But all new pages submitted for moderation will use the new moderation workflow described by this RFC.

Author: I'm thinking we could allow both to work side-by-side as it's not an exact replacement for the old workflow.

# RFC 45: Workflow

* RFC: 45
* Authors: Karl Hobley
* Created: 2019-11-12
* Last Modified: 2019-11-12

## Abstract

This RFC proposes how we can implement multi-stage configurable workflow in Wagtail core.

## Overview

### Workflows

Workflows are configured in Wagtail by an administrator. Each workflow has a name and a sequence of steps.

Workflows are triggered when a user clicks the "Submit for Review" action in the page editor. This creates a "workflow state" object and starts the first step in the workflow.

> The steps in the workflow must be completed in the same order as defined by the administrator. When the workflow is completed, the page is published.

Reviewer 3: I think it would also be good to describe what happens when a task gets done with a rejection, and the workflow doesn’t complete. My understanding from our team meeting is that this would lead to the workflow restarting from the first task?

Reviewer 3: Side question – has there been any discussion of being able to use those workflows with other models than pages, e.g. snippets, images, documents? I’ve worked on projects where Wagtail not having any concept of moderation or drafts for these was a blocker.

Author: > I think it would also be good to describe what happens when a task gets done with a rejection, and the workflow doesn’t complete. My understanding from our team meeting is that this would lead to the workflow restarting from the first task?

This logic would actually be dependant on the ``Task`` class. In the default workflow, the workflow would be cancelled and the submitter would be notified, similar to the way Wagtail works now. I think this would be the most common behaviour so we could implement it this way in the base ``Task`` class but allow it to be customised (for The Motley fool, editing the page will be implemented as a workflow task, so any rejection would return to this task rather instead of cancelling the workflow).

Author: > 

Side question – has there been any discussion of being able to use those workflows with other models than pages, e.g. snippets, images, documents? I’ve worked on projects where Wagtail not having any concept of moderation or drafts for these was a blocker.

We haven't discussed this yet, and the feature described in this RFC is dependant on revisions so it might be difficult to implement this for images/snippets/documents in a similar way.

Reviewer 3: Yes, it’s a limitation of Wagtail to only have an audit trail for page models. It doesn’t seem too worthwhile to cover this in the RFC then.

> Wagtail finds the workflow to use from the closest ancestor page that has a workflow associated with it. Out of the box, the "Root" page in Wagtail has the "Default" workflow linked with it, so this will be used in any section that hasn't got a more specific workflow assigned to it.

Reviewer 3: I (vaguely) remember people saying that it was somewhat annoying for Wagtail to have default migrations that add a Site instance (and a homepage?). Are there any reasons to be concerned about having a default workflow?

Author: I think this probably comes from the initial template including this migration. This migration will be in Wagtail core itself so this should be OK.

The purpose of creating a "Default" workflow is to replicate the current behaviour as closely as possible so we don't cause much breakage, I think most users would appreciate that.

The "Submit for Review" button is not shown when there isn't a workflow configured.

### Tasks

Each step in a workflow has a task that must be completed in order to go to the next step. Tasks are defined separately to workflows, allowing them to be to shared.

Examples of tasks include reviews by members of specific user groups, examples for this could be "Editorial Review" or "Legal Review".

> Different task types can be defined by developers using models. This allows custom tasks to be defined. For example, this could be used to implement automated grammer and spelling checking.

Reviewer: ```suggestion
Different task types can be defined by developers using models. This allows custom tasks to be defined. For example, this could be used to implement automated grammar and spelling checks.
```

Author: I honestly didn't make this mistake on purpose!

When a task starts, it creates a "task state" object. This object tracks the current progress for completing a task in a workflow for an individual revision. The "task state" object may also be customised
for each task type allowing fields like "review comment" to be implemented, for example.

> If the page has changed while a review is taking place, the workflow will return to the first step of the workflow and any "in progress" tasks will be cancelled. "task state" objects from previous reviews of the page would be available for uses such as generating diffs from the last reviewed version, they are kept indefinitely and can be viewed by users in the "Workflow history" interface.

Reviewer: ```suggestion
If the page changes while a step task is taking place, the workflow will return to the first step of the workflow and any "in progress" tasks will be cancelled. Any "task state" objects from previous steps will be kept indefinitely, and will be available for uses such as generating diffs between versions. All "task state" objects can be viewed by users in the "Workflow history" interface.
```

#### ``get_action_menu_items(workflow_state, task_state)``

This returns a list of ``ActionMenuItem`` instances to replace the defaults. This could be used to implement "Accept" and "Reject" actions.

#### ``on_action(workflow_state, task_state, action_name)``

If the edit view receives a ``POST`` request with a field called ``wagtail-moderation-action``, then this method is called
with the action name passed through.

#### ``user_can_access_editor(workflow_state, task_state, user)``

Returns ``True`` if the user is allowed to access the page editor. By default, this returns ``False``.

If this returns ``False``, Wagtail will fall back to querying ``UserPagePermissionsProxy`` so this can't be used to revoke editor permissions for a user that would usually have access to this.
This is to prevent broken links in the admin as areas of the admin which link to the edit view (such as the explorer) won't be aware of workflows.

To prevent users from editing pages while the task is active, use ``page_locked_for_user`` instead.

Also, this does not make the page appear in the explorer if the user wouldn't normally have permission to edit the page.

> #### ``page_locked_for_user(workflow_state, task_state, user)``

Reviewer: I wonder if prefixing these with `is_` will make their intended purpose clearer.

Author: Added this, thanks!

#### Adding/removing/reordering tasks when they are in use

If an admin adds a new task or reorders tasks in the workflow management interface, the current task always needs to be completed before going back to any task that was just added, even if the new task is before the current task.

For example, say we have the following workflow configured:

 Editor Approval => Clinical Approval

If a new "Grammar Check" task is added in between, any pages that are currently on "Clinical Approval" will move to "Grammar Check" after they complete "Clinical Approval", even though "Clinical Approval" is last.

If a task is removed from a workflow, any pages that are currently on that task will have to either complete it or cancel the workflow and start again.

Administrators can't permanently delete tasks if any page is currently on that task.

## UI Changes

### Page editor

When the user visits the page editor, the view checks for an active workflow on the page. If there is, certain behaviours can be overridden by the implementation of the current task in the workflow:

>  - What action buttons appear at the bottom

Reviewer 3: This feels very minor but just in case – there are also keyboard shortcuts that allow a user to save a draft and preview a page. We might want to make sure these are conditional to the corresponding buttons being there.

Author: Wagtail already allows developers to customise this menu, so this might already be an issue / has been solved.

If the edit view receives a ``POST`` request with a field called ``wagtail-moderation-action``, then this method is called
with the action name passed through.

#### ``user_can_access_editor(workflow_state, task_state, user)``

Returns ``True`` if the user is allowed to access the page editor. By default, this returns ``False``.

If this returns ``False``, Wagtail will fall back to querying ``UserPagePermissionsProxy`` so this can't be used to revoke editor permissions for a user that would usually have access to this.
This is to prevent broken links in the admin as areas of the admin which link to the edit view (such as the explorer) won't be aware of workflows.

To prevent users from editing pages while the task is active, use ``page_locked_for_user`` instead.

Also, this does not make the page appear in the explorer if the user wouldn't normally have permission to edit the page.

#### ``is_page_locked_for_user(workflow_state, task_state, user)``

Returns ``True`` if the page should be locked for the specified user. By default, this returns ``False``.

Note that this method won't unlock the page if it is locked to a specific user using Wagtail's edit lock feature.

> #### ``is_page_lockable_by_user(workflow_state, task_state, user)``

Author: Just realised this naming is inconsistent with `user_can_access_editor` so will probably rename to `user_can_lock` and `user_can_unlock` instead
